src/main/scala/com.snowplowanalytics.s3/loader/KinesisS3Emitter.scala

```diff
@@ -108,3 +126,65 @@ class KinesisS3Emitter(
   override def fail(records: java.util.List[EmitterInput]): Unit =
     s3Emitter.sendFailures(records)
 }
+
+object KinesisS3Emitter {
+
+  /** Type of row which determined according to schema of self describing data */
+  sealed trait RowType extends Product with Serializable {
+    def getName: String
+  }
+
+  object RowType {
+    /** Represents cases where row type could not be determined
+      * since either row is not valid json or it is not self
+      * describing json
+      */
+    case class PartitionError(errorDir: String) extends RowType {
+      override def getName: String = errorDir
+    }
+
+    /** Represents cases where type of row can be determined successfully
+      * e.g. does have proper schema key
+      */
+    case class SelfDescribing(rowType: String) extends RowType {
+      override def getName: String = rowType
+    }
+
+    case object UnexpectedError extends RowType {
+      override def getName: String = "unexpected_error"
+    }
+  }
+
+  val dateFormat = new SimpleDateFormat("yyyy-MM-dd");
+
+  /**
+    * Determines the filename in S3, which is the corresponding
+    * Kinesis sequence range of records in the file.
+    */
+  private def getBaseFilename(firstSeq: String, lastSeq: String, prefix: Option[String] = None): String =
+    prefix.map(p => if (p.isEmpty) "" else p + "/").getOrElse("") + dateFormat.format(Calendar.getInstance().getTime()) +
+      "-" + firstSeq + "-" + lastSeq
+
+  /**
+    * Assume records are self describing data and group them according
+    * to their schema key. Put records which are not self describing data
+    * to under "old bad row type".
+    */
+  private[loader] def partitionWithSchemaKey(records: List[EmitterInput], errorDir: String) = {
+    records.groupBy {
+      case Success(byteRecord) =>
+        val strRecord = new String(byteRecord, "UTF-8")
+        Try(parse(strRecord)) match {
+          case TrySuccess(e) =>
+            val json = parse(strRecord)
+            val schemaKey = SchemaKey.extract(json)
+            schemaKey.fold(
+              e => RowType.PartitionError(errorDir),
+              k => RowType.SelfDescribing(s"${k.vendor}.${k.name}")
```

Comment:
Did we discuess something this output dir? Why there's no model?

Reply:
I was planning to ask about it. So, do you think we should add revision-addition too or only model is enough ?

Reply:
I think it should be `vendor/name/format/model`, but worth to double-check with @yalisassoon